File: lua/vimade/style/include.lua
local M = {}

local COMPAT = require('vimade.util.compat')
local CONDITION = require('vimade.style.value.condition')
local GLOBALS

local include_key_reducer = require('vimade.util.key_reducer')()

M.__init = function (args)
  GLOBALS = args.GLOBALS
end

-- @param config = {
--  value= {'Folded', 'VertSplit', 'Normal', ...}, # list of names that should be skipped on the style array
--  style = {Fade(0.4)} # style to run on all names that aren included
--}
M.Include = function(config)
  local result = {}
  local _condition = config.condition or CONDITION.INACTIVE
  result.tick = config.tick
  result.attach = function (win)
    local condition = _condition
    local names = config.value
    local children = {}
    for i, s in ipairs(config.style) do
      table.insert(children, s.attach(win))
    end
    local include = {}
    local include_key = ''
    local style = {}
    style.win = win
    style._condition = _condition
    style._animating = false
    style.before = function (win, state)
      if type(_condition) == 'function' then
        condition = _condition(style, state)
      end
      if condition == false then
        return
      end
      local input

      if type(names) == 'function' then
        input = names(win)
      else
        input = names
      end
      if type(input) == 'string' then
        input = {input}
      end

      include_key = include_key_reducer.reduce_ipairs(input)

      include = {}
      for i, name in ipairs(input) do
        include[name] = true
      end

      for i, s in ipairs(children) do
        s.before(win, state)
      end
    end
    style.key = function (win, state)
      if condition == false then
        return ''
      end
      local key ='I-' .. include_key .. '('
      local s_active = 0
      for j, s in ipairs(children) do
        if j ~= 0 then
          key = key .. ','
        end
        local s_key = s.key(win, state)
        key = key .. s_key
        if string.len(s_key) > 0 then
          s_active = s_active + 1
        end
      end
      if s_active == 0 then
        return ''
      end
      key = key .. ')'
      return key
    end
    style.modify = function (highlights, to_hl)
      if condition == false then
        return
      end
      local highlights_for_children = {}
      for name, _ in pairs(include) do
        highlights_for_children[name] = highlights[name]
      end
      for i, s in ipairs(children) do
        s.modify(highlights_for_children, to_hl)
      end
      for name, value in pairs(highlights_for_children) do
        highlights[name] = value
      end
    end
    return style
  end
  -- value is not exposed here, no use-case currently
  return result
end
return M

File: lib/vimade/style/include.py
import sys
M = sys.modules[__name__]

from vimade.style.value import condition as CONDITION

GLOBALS = None

def __init(args):
  global GLOBALS
  GLOBALS = args['GLOBALS']
M.__init = __init


M.include_names = {}
M.include_id = 1
# @param config = {
#  'value': ['Folded', 'VertSplit', 'Normal', ...], # list of names that should be skipped on the style array
#  'style': [Fade(0.4)] # style to run on all names that are include
# }
class Include():
  def __init__(self, **kwargs):
    _condition = kwargs.get('condition')
    _condition = _condition if _condition != None else CONDITION.INACTIVE
    _names = kwargs.get('value', [])
    self.tick = kwargs.get('tick')
    class __Include():
      def __init__(self, win, state):
        self.win = win
        self.condition = None
        self._condition = _condition
        self.names = []
        self.style = [s.attach(win, state) for s in kwargs.get('style', [])]
        self.include = {}
        self._animating = False
      def before(self, win, state):
        self.condition = _condition(self, state) if callable(_condition) else _condition
        if self.condition == False:
          return
        names = self.names = _names(self, state) if callable(_names) else _names
        self.include = include = {}
        if type(names) == str:
          names = [names]
        for name in names:
          name_id = M.include_names.get(name)
          if not name_id:
            M.include_names[name] = name_id = str(M.include_id)
            M.include_id += 1
          include[name] = name_id
        for s in self.style:
          s.before(win, state)
      def key(self, win, state):
        if self.condition == False:
          return ''
        style_keys = [s.key(win, state) for j, s in enumerate(self.style)]
        style_key = ','.join([s for s in style_keys if s])
        if not style_key:
          return ''
        return 'I-' + ','.join(self.include.values()) + '(' \
            + style_key + ')'
      def modify(self, highlights, to_hl):
        if self.condition == False:
          return
        highlights_for_children = {}
        for name in self.include:
          if name in highlights:
            highlights_for_children[name] = highlights[name]
        for s in self.style:
          s.modify(highlights_for_children, to_hl)
        for (name, value) in highlights_for_children.items():
          highlights[name] = value
    self.attach = __Include
